File: TUnit.TestProject/DependencyInjectionClassConstructor.cs
using System.Diagnostics.CodeAnalysis;
using Microsoft.Extensions.DependencyInjection;
using TUnit.Core.Interfaces;

namespace TUnit.TestProject;

public class DependencyInjectionClassConstructor : IClassConstructor
{
    private readonly IServiceProvider _serviceProvider = CreateServiceProvider();

    public Task<object> Create([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type type, ClassConstructorMetadata classConstructorMetadata)
    {
        var instance = ActivatorUtilities.GetServiceOrCreateInstance(_serviceProvider, type);
        return Task.FromResult(instance);
    }

    private static IServiceProvider CreateServiceProvider()
    {
        var services = new ServiceCollection();
        services.AddTransient<DummyReferenceTypeClass>();
        return services.BuildServiceProvider();
    }
}

File: TUnit.TestProject.VB.NET/DependencyInjectionClassConstructor.vb
Imports System
Imports System.Threading.Tasks
Imports Microsoft.Extensions.DependencyInjection
Imports TUnit.Core
Imports TUnit.Core.Interfaces

Public Class DependencyInjectionClassConstructor
    Implements IClassConstructor

    Private ReadOnly _serviceProvider As IServiceProvider = CreateServiceProvider()

    Public Function Create(type As Type, classConstructorMetadata As ClassConstructorMetadata) As Task(Of Object) Implements IClassConstructor.Create
        Dim instance = ActivatorUtilities.GetServiceOrCreateInstance(_serviceProvider, type)
        Return Task.FromResult(instance)
    End Function

    Private Shared Function CreateServiceProvider() As IServiceProvider
        Dim services = New ServiceCollection()
        services.AddTransient(Of DummyReferenceTypeClass)()
        Return services.BuildServiceProvider()
    End Function

End Class
